### PR TITLE
Add image sharing to application

### LINKED ISSUE #11: Share Image
#### As a user I want to share an image with someone via email.

__Story__:

This application is called Image Sharer. Sure, you can share images by adding
them to our web service, but how will people who do not know about this web
service be able to see those images? In phone calls with our customers we have
discovered that many of our users are manually emailing their friends links to
the images. That is awesome! We can make things more efficent for our users by
directly supporting sharing images through email.

__Acceptance criteria__:
- [x] There is a link to share an image.
- [x] After clicking on this link, a bootstrap modal appears with a form to
  send the email.
- [x] The form has fields for the recipient's email and an optional message.
- [x] The email we send has an HTML part and a text part.
- [x] The email contains the image, the custom message if provided, and a link
  inviting the recipient to check out our application.
- [x] The modal closes and a success message is shown if the email was sent
  successfully.
- [x] The user is alerted of errors if the operation was unsuccessful.

__Note__: This story is large, so it makes sense to break the implementation
down into more manageable pieces (a la elephant carpaccio). Please start by
implementing a basic sharing feature that uses a new page (not a modal), and
open a PR for that. Then, once we've verified the simplified version is
working, refactor it to use a modal, and open a second PR for that. The
[video](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/basics-of-bootstrap-modals-js-and-ajax)
provides some background and code examples that will be really helpful for the
second PR, so it's worth watching (and following along) before you begin that
part.

__Discussion topics__:
- [x] How could we develop the front end and back end separately? (And why
  might that be a good idea?)
- [x] How should we implement a non-CRUD action?
- [x] What are the different types of errors we should handle?
- [x] What is AJAX? What libraries can help us add AJAX to our Rails app?
- [x] How should we go about sending emails from our Rails app?
- [x] How could we show a preview of the image being shared in the modal?
- [x] What are the implications of allowing users to send emails from our app?

__Reference material__:
- http://guides.rubyonrails.org/action_mailer_basics.html
- https://elements.heroku.com/addons/sendgrid
- https://sendgrid.com/docs/Integrate/Frameworks/rubyonrails.html
- https://github.com/rails/jquery-ujs/wiki
- http://v4-alpha.getbootstrap.com/components/modal
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/basics-of-bootstrap-modals-js-and-ajax

__Dependencies__:
- Image Index
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by mathewbaltes**
Closes #11

### CODE REVIEW

**Comment by lsq921224** [comment]
I don't think the sharing functionality works on the review app. Can you first make sure it works properly?

File: app/assets/javascripts/images/share-modal.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    $('#share-modal').modal('show');
+    e.preventDefault();
+  };
+
+  hide() {
+    $('#share-modal').modal('hide');
+  }
+
+  init() {
+    var self = this;
```
Comment (lsq921224) [inline app/assets/javascripts/images/share-modal.js]: first of all.. `let` is preferred over `var` in es6.. secondly.. I don't see a need for the reference `this` here. Even if we do, we can use arrow functions of ES6 to prevent reference `this`

File: app/assets/javascripts/images/share-modal.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+class ShareModal {
+
```
Comment (lsq921224) [inline app/assets/javascripts/images/share-modal.js]: extra empty line here

File: app/assets/javascripts/images/main.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+const Main = {
+  init: () => {
+    ShareModal.init();
+  }
+};
+
+// Init the main function for image
+Main.init();
+
```
Comment (lsq921224) [inline app/assets/javascripts/images/main.js]: extra empty new line

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -36,9 +36,27 @@ def show
     redirect_to root_path
   end
 
+  def share
+    @email = Email.new(email_params)
+    if @email.invalid?
+      render status: 400, json: @email.to_json
```
Comment (lsq921224) [inline app/controllers/images_controller.rb]: The more appropriate response for invalid input is 422, Unprocessable Entity, and it's better to use symbol instead of number code here. e.g. `status: :unprocessable_entity`

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -36,9 +36,27 @@ def show
...
+    else
+      UserMailer.share_email(@email).deliver_now
+      # Renders 204 by default
```
Comment (lsq921224) [inline app/controllers/images_controller.rb]: Normally we return a 200 if operation succeeds, you could do `head :ok` here

File: app/assets/javascripts/images/share-modal.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    $('#share-modal-form').on('ajax:success', () => {
+      alert('Successfully shared image!');
+      self.hide();
+    });
+    $('#share-modal-form').on('ajax:error', () => {
+      alert('Failed to send image');
+      self.hide();
+    });
+  };
+};
```
Comment (lsq921224) [inline app/assets/javascripts/images/share-modal.js]: extra semicolon

File: test/page_objects/images/share_modal.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+module PageObjects
+  module Images
+    class ShareModal < PageObjects::Document
```
Comment (lsq921224) [inline test/page_objects/images/share_modal.rb]: Modal is not a page, this should be `AePageObjects::Element`

File: test/page_objects/images/image_card.rb [hunk 8]
```diff
@@ -9,6 +9,11 @@ def tags
         node.all('.js-tag').map(&:text)
       end
 
+      def click_share!
+        node.find('.js-share-image').click
+        window.change_to(ShareModal)
```
Comment (lsq921224) [inline test/page_objects/images/image_card.rb]: Since we are poping up a modal, we are not changing to another page

File: config/environments/production.rb [hunk 9]
```diff
@@ -83,4 +83,14 @@
 
   # Do not dump schema after migrations.
   config.active_record.dump_schema_after_migration = false
+
```
Comment (lsq921224) [inline config/environments/production.rb]: as mentioned before, this change and changes in `development.rb` should belong to one commit

File: app/views/images/show.html.erb [hunk 10]
```diff
@@ -12,7 +12,7 @@
 </div>
 <div class="row py-1 justify-content-center">
   <div class="col text-center">
-    <%= render partial: 'images/tag_list', locals: {tag_list: @image.tag_list} %>
+    <%= render partial: 'tag_list', locals: { tag_list: @image.tag_list} %>
```
Comment (lsq921224) [inline app/views/images/show.html.erb]: the change in https://github.com/AppFolioOnboarding/image-sharer-mathewbaltes/pull/32/files#diff-54003fc1b9e144a95e605d0326f6507aR25 and this line are refactoring, they should not belong to this commit

File: test/flows/images_crud_test.rb [hunk 11]
```diff
@@ -86,4 +86,48 @@ class ImagesCrudTest < FlowTestCase
...
+    share_modal.send_email!(
+      address: 'a@a.com',
+      message: 'blah'
+    ) do |confirm_dialog|
+      assert_equal 'Successfully shared image!', confirm_dialog.text
+      confirm_dialog.dismiss
+    end
+  end
+
+  test 'share incorrect image' do
```
Comment (lsq921224) [inline test/flows/images_crud_test.rb]: Another common case is sharing a nonexistent image, can we add a test for that?

File: test/flows/images_crud_test.rb [hunk 12]
```diff
@@ -86,4 +86,48 @@ class ImagesCrudTest < FlowTestCase
     images_index_page = images_index_page.clear_tag_filter!
     assert_equal 3, images_index_page.images.count
   end
+
```
Comment (lsq921224) [inline test/flows/images_crud_test.rb]: These two test cases are not CRUD, they should not be in this file.

File: test/controllers/images_controller_test.rb [hunk 13]
```diff
@@ -149,6 +149,24 @@ def setup
...
+    link_id = Image.create(url: 'http://test.com/', tag_list: 'test').id
+    post share_image_path(link_id), params: { email: { address: 'blah' } }
+
+    assert_response :bad_request
+    assert_equal 'blah', ActiveSupport::JSON.decode(@response.body)['address']
+    assert_equal image_url(link_id), ActiveSupport::JSON.decode(@response.body)['url']
+    assert_equal 'http://test.com/', ActiveSupport::JSON.decode(@response.body)['img']
+    assert_empty ActiveSupport::JSON.decode(@response.body)['message']
+  end
+
```
Comment (lsq921224) [inline test/controllers/images_controller_test.rb]: Another common case is sharing a nonexistent image, can we add a test for that?

File: test/models/email_model_test.rb [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+  test 'verify email with valid address is valid' do
+    email = Email.new('address': 'a@a.com')
+    assert_equal true, email.valid?
+    assert_empty email.errors
+  end
+
+  test 'verify email with invalid address is valid' do
+    email = Email.new('address': 'a.com')
+    assert_equal false, email.valid?
+    assert_not_empty email.errors
```
Comment (mathewbaltes) [inline test/models/email_model_test.rb]: Should probably assert the other fields as well.

File: test/page_objects/images/show_page.rb [hunk 15]
```diff
@@ -25,6 +25,17 @@ def delete_and_confirm!
       def go_back_to_index!
         IndexPage.visit
       end
+
+      def click_share!
```
Comment (lsq921224) [inline test/page_objects/images/show_page.rb]: Since this is not opening a new page, it shouldn't have the bang in the method name

File: test/page_objects/images/share_modal.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+        node.find('#modal-notification > #flash-error', visible: true)
+      end
+
+      def send_email!(address: nil, message: nil)
+        self.address.set(address) if address.present?
+        self.message.set(message) if message.present?
+        node.click_button('Share image')
+      end
+
+      def hide!
```
Comment (lsq921224) [inline test/page_objects/images/share_modal.rb]: Since this is not opening a new page, it shouldn't have the bang in the method name

File: test/page_objects/images/share_modal.rb [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+      form_for :email do
+        element :address
+        element :message
+      end
+
+      def form_error?
+        node.find('#modal-notification > #flash-error', visible: true)
+      end
+
+      def send_email!(address: nil, message: nil)
```
Comment (lsq921224) [inline test/page_objects/images/share_modal.rb]: Is this opening a new page?

File: test/page_objects/images/image_card.rb [hunk 18]
```diff
@@ -9,6 +9,10 @@ def tags
         node.all('.js-tag').map(&:text)
       end
 
+      def click_share!
```
Comment (lsq921224) [inline test/page_objects/images/image_card.rb]: Is this opening a new page?

File: app/views/images/index.html.erb [hunk 19]
```diff
@@ -9,17 +9,18 @@
     <% if flash[:error] %>
       <%= content_tag(:div, flash[:error], id: 'flash-error', class: 'alert alert-danger') %>
     <% end %>
+    <div id="index-notification"></div>
```
Comment (lsq921224) [inline app/views/images/index.html.erb]: This div, if I understand correctly, is the div that contains the flash message. If so, then all the 
```
       <%= content_tag(:div, flash[:error], id: 'flash-error', class: 'alert alert-danger') %>
```
should be put inside this div. Otherwise we might see two flash messages in some cases

File: app/views/images/show.html.erb [hunk 20]
```diff
@@ -1,8 +1,10 @@
+<%= render partial: 'share_modal' %>
 <div class="row justify-content-center">
   <div class="col-6 text-center">
     <% if flash[:success] %>
         <%= content_tag(:div, flash[:success], id: 'flash-success', class: 'alert alert-success') %>
     <% end %>
+    <div id="index-notification"></div>
```
Comment (lsq921224) [inline app/views/images/show.html.erb]: Same comment as before. besides, flash messages should better be put in the layout as a partial, so that you don't need to do this in every erbs. e.g
```
    <%= render partial: 'layouts/header' %>
     ....
     <div id="flash_messages">
        <%= render partial: 'layouts/flash_messages' %>
      </div>
      <%= yield %>
     ...
```

File: app/assets/javascripts/images/share-modal.js [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+    $('#share-modal').modal('show');
+    e.preventDefault();
+  };
+
+  hide() {
+    $('#share-modal').modal('hide');
+  }
+
+  success() {
+    $('#index-notification').html(
```
Comment (lsq921224) [inline app/assets/javascripts/images/share-modal.js]: Do you think there are some duplications in this code?
```
$('#index-notification').html(
        '<div id="flash-error" class="alert alert-danger">Sorry, the image no longer exists and cannot be shared.</div>'
 );
```
```
$('#modal-notification').html(
        '<div id="flash-error" class="alert alert-danger">Please insert a valid email.</div>'
  );
```
and 
```
 $('#index-notification').html(
'<div id="flash-success" class="alert alert-success">Thank you for sharing the image with your friend!</div>'
   );
```
are very similar, and there should be a specific method that we can call to handle the flash messages

File: app/controllers/images_controller.rb [hunk 22]
```diff
@@ -36,9 +36,32 @@ def show
...
+    @image = Image.find(params[:id])
+
+    if @email.invalid?
+      render status: :unprocessable_entity, json: @email.to_json
+    else
+      UserMailer.share_email(@email).deliver_now
+      head :ok
+    end
+
+  rescue ActiveRecord::RecordNotFound
```
Comment (lsq921224) [inline app/controllers/images_controller.rb]: `rescue` should be used to handle _exceptional_ behavior. In this case, we don't need to use `rescue` if we just use `Image.find_by(id: params[:id])` instead. That will return either the image or `nil`, so you can just write an `if` `else` statement to handle the cases of whether the image exists or not.

File: test/page_objects/images/image_card.rb (hunk 18 above)
Comment (mathewbaltes) [inline test/page_objects/images/image_card.rb]: No, refactored to remove !

File: app/assets/javascripts/images/share-modal.js [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    $('#email_message').val('');
+    $('#share-modal').modal('show');
+    e.preventDefault();
+  };
+
+  hide() {
+    $('#share-modal').modal('hide');
+  }
+
+  flashMessage(id, type, message) {
```
Comment (lsq921224) [inline app/assets/javascripts/images/share-modal.js]: if we look at what this method does, we will find that this is a "general purpose" function, therefore we can move this function out to anther file, e.g `shared/flash_message.js`, export it and then import it here

File: app/views/images/show.html.erb [hunk 24]
```diff
@@ -1,8 +1,7 @@
+<%= render partial: 'share_modal' %>
 <div class="row justify-content-center">
   <div class="col-6 text-center">
-    <% if flash[:success] %>
-        <%= content_tag(:div, flash[:success], id: 'flash-success', class: 'alert alert-success') %>
-    <% end %>
+    <%= render partial: 'layouts/flash_messages' %>
```
Comment (lsq921224) [inline app/views/images/show.html.erb]: this is much better, but as I mentioned, we can move this to the template file (https://github.com/AppFolioOnboarding/image-sharer-mathewbaltes/blob/mb_issue11/app/views/layouts/application.html.erb#L12), since every page will need flash messages most likely
